### ISSUE TITLE
cannot import 'tap' after lockdown (source-map-support)

### ISSUE THREAD

**Issue by warner** [Agoric/SES-shim#367]
This is an ongoing compatibility report with the ['tap' module](https://github.com/tapjs/node-tap). `tap` provides three parts of a unit-testing environment: the runner (a command-line tool that looks for test programs, runs them, and collects/formats the results), the test function (the `tap.test(NAME, testfn)` thing to execute inside your test programs), and a librart of assertions (the `t.equal(x,y)` inside those test functions). We use `tap` in SES-shim for all unit tests.

I'm looking at the following criteria:

* 1: `import tap from 'tap'` works
* 2: failed assertions show position of, and stack trace to, the assertion that failed
* 3: unexpected exceptions show the Error message, location and stack of the `throw`, and flunk the test
* 4: `t.throws` accepts the right exception
* 5: `t.throws` displays Error message details of wrong exceptions

I'm testing against current SES trunk (bfe610fe5fe7afd31659fc8e40ee6e77d622e264 , somewhat after SES-0.8.0, includes @erights Error-constructor taming).

## import `tap` before `lockdown()` in full-safe-mode

* 1: import works without complaint
* 2: failed assertions show difference between expected and actual, but the position and stack trace are missing
* 3: unexpected exceptions show Error, stack trace, and flunk the test. However top stack frame and location point into the taming wrapper (`tame-global-error-object.js:61:14`). The second stack frame points to the actual `throw`.
* 4: `t.throws` accepts the right exception
* 5: `t.throws`displays Error message, but not location+stack of wrong exceptions

## import `tap` before `lockdown()` with `errorTaming: 'unsafe'`

* 1: import works without complaint
* 2: failed assertions trigger internal error: `frame.isNative is not a function`, does not display location/stack of failed assertion, terminates test early, flunks test
* 3: unexpected exceptions show Error, stack trace, and flunk the test, but top stack frame points into taming wrapper (second stack frame is the `throw`). Same as full-safe-mode.
* 4: `t.throws` accepts the right exception
* 5: `t.throws` on the wrong exception triggers the internal `frame.isNative` error

## import `tap` after `lockdown()` in full-safe mode

* 1: import works
* 2: failed assertions display expected-vs-actual difference, but position and stack trace are missing
* 3: unexpected exceptions show extra stack frame as above
* 4: `t.throws` accepts the right exception
* 5: `t.throws` omits location+stack of wrong exceptions, as above

## import `tap` after `lockdown()` with `errorTaming: 'unsafe'`

* 1: import works, but any call to the `tap.test` function triggers the internal `frame.isNative is not a function` error before any of the assertions are run, and the test function is flunked

## Notes

`tap` works in the SES unit tests because it is imported before `lockdown()`. In agoric-sdk, we use the `import 'install-ses'` pattern, in which we call `lockdown()` as a side-effect of importing a module, so that all other code is loaded in a SES environment.

`tap` uses a module named [source-map-support](https://www.npmjs.com/package/source-map-support) which, on Node, modifies `Error` to use any `sourceMap` it can find to produce stack frames with more accurate filename/line-number data. It [re-implements](https://github.com/evanw/node-source-map-support/blob/d29e9c81527346b6ec385f7ba27a7a1c487b69c7/source-map-support.js#L272) v8's `CallSiteToString`, which calls `callSite.isNative()` without guarding against that function being missing. It also wraps each `callSite` with [another function](https://github.com/evanw/node-source-map-support/blob/d29e9c81527346b6ec385f7ba27a7a1c487b69c7/source-map-support.js#L348) that returns the sourcemap-derived locations, and that wrapper uses `.isNative()` again. It [replaces](https://github.com/evanw/node-source-map-support/blob/d29e9c81527346b6ec385f7ba27a7a1c487b69c7/source-map-support.js#L563) `Error.prepareStackTrace` with [a version](https://github.com/evanw/node-source-map-support/blob/d29e9c81527346b6ec385f7ba27a7a1c487b69c7/source-map-support.js#L413) that wraps each `callSite`. It also has an option to replace `process.emit` to catch uncaught exceptions.

`source-map-support` has 20M weekly downloads, and 4k immediate downstream dependent packages.

**Comment by warner** [comment]
One problem is that we aren't exposing `isNative()` on our wrapped `callSite`. I *think* that ought to be safe.

I tried adding that, but the problem remained. I noticed that `source-map-support` doesn't use `callSite.isNative()` directly. Instead, it builds its own `callSite` wrapper like this:

```js
function cloneCallSite(frame) {
  var object = {};
  Object.getOwnPropertyNames(Object.getPrototypeOf(frame)).forEach(function(name) {
    object[name] = /^(?:is|get)/.test(name) ? function() { return frame[name].call(frame); } : frame[name];
  });
  object.toString = CallSiteToString;
  return object;
}
```

And then inside their `CallSiteToString` function, they use `this.isNative()` (maybe because this code was copied out of the v8 internals?):

```js
function CallSiteToString() {
  var fileLocation = "";
  if (this.isNative()) {
    fileLocation = "native";
    ..
```

It looks like they're depending upon a funny property of the v8 `callSite` object, which is that the various accessor-like functions are not properties of the `callSite` itself, but rather on that object's prototype. Our `callSite` wrapper puts them directly on the object, so `source-map-support` doesn't copy anything onto their wrapper.

I grabbed a list of all the properties of a v8 `callSite`'s prototype, and found two (`getPosition` and `getScriptNameOrSourceURL`) that were not listed in the [v8 docs](https://v8.dev/docs/stack-trace-api):

```js
function prep(error, sst) {
  const top = sst[0];
  console.log(`names:`, Object.getOwnPropertyNames(top));
  console.log(`proto names:`, Object.getOwnPropertyNames(Object.getPrototypeOf(top)));
  return 'fake stack';
}
Error.prepareStackTrace = prep;
Error().stack;
```

```
names: []
proto names: [
  'constructor',
  'getColumnNumber',
  'getEvalOrigin',
  'getFileName',
  'getFunction',
  'getFunctionName',
  'getLineNumber',
  'getMethodName',
  'getPosition',
  'getPromiseIndex',
  'getScriptNameOrSourceURL',
  'getThis',
  'getTypeName',
  'isAsync',
  'isConstructor',
  'isEval',
  'isNative',
  'isPromiseAll',
  'isToplevel',
  'toString'
]
```

`getPosition` appears to return an integer with the character offset of the `Error` invocation, so it contains roughly the same information as `getLineNumber` and `getColumnNumber`. `source-map-support` does not use `getPosition`, nor do I see any users in the agoric-sdk `node_modules/` tree.

`getScriptNameOrSourceURL` appears to return a `file:` URL (in my tests), and `source-map-support` uses it.

`source-map-support` also calls `this.isToplevel()`, which is misspelled as `isTopLevel` (capital-L) in `safeV8CallSiteMethodNames`.

I'm checking to see if I can modify `tame-global-error-object.js` enough to let this work. I'll be filing PRs when I've got something working.

**Comment by warner** [comment]
By enabling `isNative`, `getPosition`, `getScriptNameOrSourceURL` to the safe-list, fixing the spelling of `isToplevel`, and moving the properties to a prototype, I get:

## import tap after `lockdown()` with `errorTaming: 'unsafe'`

1: import works
2: failed assertions display expected-vs-actual difference, correct position, correct stack
3: unexpected exceptions show extra stack frame as above
4: t.throws accepts the right exception
5: t.throws on the wrong exception shows the correct location+stack

## import tap after `lockdown()` with full safe-mode

1: import works
2: failed assertions omit location+stack
3: unexpected exceptions show extra stack frame as above
4: t.throws accepts the right exception
5: t.throws on the wrong exception omits location+stack

**Comment by warner** [comment]
Ok #369 landed the addition of those four properties, and moved them to a prototype. I added a test which imports `tap` after doing `lockdown()`, and does at least a few basic operations. I can't automate the testing of how `tap` responds to failed assertions, but I've manually checked those and they all look as good as we can get right now.

Programs which just look at `.stack`, without trying to get fancy with `.prepareStackTrace`, are rewarded for their humility by getting a stack in both `safe` and `unsafe` mode. Programs like `tap` which reach for more wind up getting nothing in certain circumstances (failed assertions). This might be worth further investigation, but maybe `tap` is not prepared for it's `.prepareStackTrace` to fail.

I've added #370 to discuss the unexpected extra stack frame.

closing this now
